src/http/server/response.cr

```diff
@@ -149,6 +149,26 @@ class HTTP::Server
       respond_with_status(HTTP::Status.new(status), message)
     end
 
+    # Sends a redirect to *location*.
+    #
+    # The value of *location* gets encoded with `URI.encode`.
+    #
+    # The *status* determines the HTTP status code which can be
+    # `HTTP::Status::FOUND` (`302`) for a temporary redirect or
+    # `HTTP::Status::MOVED_PERMANENTLY` (`301`) for a permanent redirect.
+    #
+    # The response gets closed.
+    #
+    # Raises `IO::Error` if the response is closed or headers were already
+    # sent.
+    def redirect(location : String | URI | Path, status : HTTP::Status = :found)
+      check_headers
+
+      self.status = status
+      headers["Location"] = URI.encode(location.to_s)
```

Comment:
A bit redundant to convert the URI to a string and then encode it again

Reply:
I don't think there is a short cut for serializing a `URI` instance directly to an URI-encoded string.

Reply:
Also `URI.encode` is deprecated, unless an IO is passed. `URI.encode_path` should be used instead.

Reply:
> `URI.encode_path` should be used instead.

No, I don't think that would be right. `encode_path` encodes reserved characters which need to be preserved in a full URI.

Reply:
The semantics of `URI.encode` which encodes any non-ASCII character as well as any special characters that have no special meaning in a URI.
Encoding only non-ASCII characters would probably be sufficient as well, though.

This might be a good use case for adding another overload of `URI.encode` which I asked for in https://github.com/crystal-lang/crystal/pull/11248#discussion_r716194817.

Reply:
> I don't think there is a short cut for serializing a URI instance directly to an URI-encoded string.

Sorry if I'm not understanding something, but why wouldn't `URI#to_s` work? Is there a spec to prove that if you pass a `URI` here the Location header gets a different value than doing `URI#to_s`?

Reply:
If a URI contains control characters they must be escaped. This could be a potential security issue, but the `HTTP::Headers` has this covered. It will still result in an error.

```cr
require "http"

headers = HTTP::Headers.new
headers["Location"] = URI.parse("http://example.com/\nLocation: http://adversary.com").to_s # ArgumentError: Header content contains invalid character '\n'
```

Reply:
Well, in any case it's probably not a big deal to do this roundtrip. We could always optimize it later if needed.